Support struct.new, struct.get and struct.set.

dbezhetskov opened the pull request:
#### 481cb6061b396e2de991a7667240e526f3787eb4
<pre>
Support struct.new, struct.get and struct.set.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243208">https://bugs.webkit.org/show_bug.cgi?id=243208</a>

Reviewed by Yusuke Suzuki.

This patch introduces basic support of all main structs operations - struct.new, struct.get and struct.set
from the Wasm GC proposal (<a href="https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#structures).">https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#structures).</a>
In this patch struct is represented as a contiguous memory space with a corresponding JS part.
All operations are implemented via runtime calls so it&apos;s not optimal for B3 and Air tiers
but optimization (replacing runtime call to a read/write by known offset) for struct.get/set will be added as a next step.

* JSTests/wasm/gc/structs.js:
(testStructDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-output.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::AirIRGenerator::emitWriteBarrier):
(JSC::Wasm::AirIRGenerator::addStructNew):
(JSC::Wasm::AirIRGenerator::addStructGet):
(JSC::Wasm::AirIRGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::B3IRGenerator::emitWriteBarrier):
(JSC::Wasm::B3IRGenerator::emitStructSet):
(JSC::Wasm::B3IRGenerator::addStructNew):
(JSC::Wasm::B3IRGenerator::addStructGet):
(JSC::Wasm::B3IRGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseStructTypeIndex):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseStructFieldIndex):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseStructTypeIndexAndFieldIndex):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseStructFieldManipulation):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addStructNew):
(JSC::Wasm::LLIntGenerator::addStructGet):
(JSC::Wasm::LLIntGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseStructType):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::StructType::StructType):
(JSC::Wasm::TypeDefinition::tryCreateStructType):
(JSC::Wasm::StructParameterTypes::translate):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::typeKindSizeInBytes):
(JSC::Wasm::StructType::getFieldOffset const):
(JSC::Wasm::StructType::getFieldOffset):
(JSC::Wasm::StructType::instancePayloadSize const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedStructSize):
(JSC::Wasm::StructType::StructType): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::toJSValue):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp: Added.
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
(JSC::JSWebAssemblyStruct::tryCreate):
(JSC::JSWebAssemblyStruct::fieldPointer const):
(JSC::JSWebAssemblyStruct::fieldPointer):
(JSC::JSWebAssemblyStruct::get const):
(JSC::JSWebAssemblyStruct::set):
(JSC::JSWebAssemblyStruct::finishCreation):
(JSC::JSWebAssemblyStruct::visitChildrenImpl):
(JSC::JSWebAssemblyStruct::destroy):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h: Added.
* Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.cpp: Added.
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::WebAssemblyStructConstructor::create):
(JSC::WebAssemblyStructConstructor::createStructure):
(JSC::WebAssemblyStructConstructor::finishCreation):
(JSC::WebAssemblyStructConstructor::WebAssemblyStructConstructor):
* Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.h: Added.
* Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.cpp: Added.
(JSC::WebAssemblyStructPrototype::create):
(JSC::WebAssemblyStructPrototype::createStructure):
(JSC::WebAssemblyStructPrototype::finishCreation):
(JSC::WebAssemblyStructPrototype::WebAssemblyStructPrototype):
* Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.h: Added.
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/255586@main">https://commits.webkit.org/255586@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e47ab90e981d88e5eb4bcf06264ca31c8f036b4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2203 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102714 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2213 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30531 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85386 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98828 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98667 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1515 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79476 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28431 "Found 1 new test failure: media/video-as-img-output-pts.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83431 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71542 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84323 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36928 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17051 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79392 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34739 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18250 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27491 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3871 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38612 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40847 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82023 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40537 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37439 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18553 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2324,6 +2327,107 @@ auto B3IRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> Pa
+{
+    const auto type = Type { TypeKind::Ref, Nullable::No, m_info.typeSignatures[typeIndex]->index() };
+    PatchpointValue* patch = m_proc.add<PatchpointValue>(toB3Type(type), origin());
+    patch->append(instanceValue(), ValueRep::reg(GPRInfo::argumentGPR0));
```
Missing FP?

dbezhetskov replied:
sorry, don't understand why should I add floating point reg here, could you please clarify it a bit?

justinmichaud replied:
This patchpoint expects the call frame as the second argument, right?

dbezhetskov replied:
right, I've added it

justinmichaud commented on Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
```diff
@@ -1886,6 +1889,39 @@ auto LLIntGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+{
+    // We have to materialize the arguments here since it might include constants or
+    // delayed moves, but the wasm_struct_new opcode expects all the arguments to be contiguous
+    // in the stack. The reason we don't call materializeConstantsAndLocals here is that
```
Can we change it to be polymorphic, or change what it expect?

dbezhetskov replied:
Yeah, we could make `materializeConstantsAndLocals` to be polymorphic but the main point here is that `materializeConstantsAndLocals` works with containers of `TypedExpression`s but here we have only container of `ExpressionType`s so that is why we can't use it here. The comment is misleading a bit, so I've shortened it.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -829,6 +830,42 @@ JSC_DEFINE_JIT_OPERATION(operationWasmRefFunc, EncodedJSValue, (Instance* instan
+    for (unsigned i = 0; i < structType.fieldCount(); ++i)
+        values[i] = arguments[i];
+
+    JSValue structValue = JSWebAssemblyStruct::create(globalObject, vm, globalObject->webAssemblyStructStructure(), Wasm::WasmStruct::create(WTFMove(structTypeDefinition), WTFMove(values)));
```
Should this just take in a global object?

dbezhetskov replied:
Thanks, I've removed a few parameters that can be obtained through global obj.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
```diff
@@ -163,6 +163,35 @@ static inline void emitThrowImpl(CCallHelpers& jit, unsigned exceptionIndex)
 }
 
+static inline void emitStructNewImpl(CCallHelpers& jit, unsigned typeIndex)
+{
```
Should we assert wasm gc is enabled?

dbezhetskov replied:
Added

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -1096,6 +1173,59 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
             return { };
         }
+        case GCOpType::StructNew: {
+            uint32_t typeIndex;
```
check if wasm gc is enabled?

dbezhetskov replied:
it isn't needed because we have a check for GCPrefix:
```
case GCPrefix: {
        WASM_PARSER_FAIL_IF(!Options::useWebAssemblyGC(), "Wasm GC is not enabled");
```

justinmichaud commented on Source/JavaScriptCore/wasm/WasmStruct.cpp:
```diff
@@ -0,0 +1,112 @@
+
+    size_t fieldOffset = 0;
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSize(fieldKind(i));
```
What does the spec say about padding? I guess that is for future optimization though, just a thought

dbezhetskov replied:
AFAIU spec doesn't say anything about padding or layout of structs.
There are plenty of optimizations that we can have here, but yeah, let's land the initial implementation first.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmStruct.cpp:
```diff
@@ -0,0 +1,112 @@
+
+void WasmStruct::set(uint32_t fieldIndex, uint64_t value)
+{
+    uint8_t* targetPointer = reinterpret_cast<uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex);
```
What is the plan for WASM gc write barriers? If a field with a reference type will just store the JSValue, don't we need a write barrier here?

dbezhetskov replied:
added for reference types

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,86 @@
+        return nullptr;
+    }
+
+    auto* structValue = new (NotNull, allocateCell<JSWebAssemblyStruct>(vm)) JSWebAssemblyStruct(vm, structure, WTFMove(nativeStruct));
```
GC enabled check?

dbezhetskov replied:
sure, I've added one

dbezhetskov commented on JSTests/wasm/gc/structs.js:
```diff
@@ -152,5 +152,616 @@ function testStructJS() {
   )
 }
 
+function testStructNew() {
```
I haven't used new `instantiate()` api because `struct.new/get/set` aren't supported there yet.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
```diff
@@ -163,6 +163,37 @@ static inline void emitThrowImpl(CCallHelpers& jit, unsigned exceptionIndex)
+    GPRReg scratch = GPRInfo::nonPreservedNonArgumentGPR0;
+    jit.loadPtr(CCallHelpers::Address(GPRInfo::argumentGPR0, Instance::offsetOfOwner()), scratch);
+    {
+        auto preciseAllocationCase = jit.branchTestPtr(CCallHelpers::NonZero, scratch, CCallHelpers::TrustedImm32(PreciseAllocation::halfAlignment));
```
Should we make a materializeInstance helper? Why isn't it pinned, or why can't we rematerialize it to the pinned register here?

justinmichaud replied:
Still confused about this

justinmichaud reviewed (commented):
Overall, everything is looking good to me.

I have no idea how to check the JSWebAssemblyStruct code, so I am going to ping someone to review that (@Constellation @saambarati @MenloDorian?).

Can we change the asserts that WASM GC is enabled to release asserts for now, just to make sure this doesn't expose any new attack surface?

saambarati commented on Source/JavaScriptCore/wasm/WasmStruct.h:
```diff
@@ -0,0 +1,91 @@
+
+class Instance;
+
+class WasmStruct final : public ThreadSafeRefCounted<WasmStruct> {
```
Can this and JSWasmStruct be merged?

saambarati replied:
Just as an FYI, there was no need for this to be reference counted. Our ability to merge these two types is a proof we wouldn't have needed this to be ref counted.

dbezhetskov replied:
Merged

saambarati commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2019-2022 Apple Inc. All rights reserved.
```
2022, not 2019

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/WasmStruct.cpp:
```diff
@@ -0,0 +1,154 @@
+
+void WasmStruct::set(JSGlobalObject* globalObject, uint32_t fieldIndex, JSValue argument)
+{
+    uint8_t* targetPointer = reinterpret_cast<uint8_t*>(this) + offsetOfTail() + getFieldOffset(fieldIndex);
```
nit: bitwise_cast is a little nicer here

dbezhetskov replied:
agree, fixed

saambarati commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.cpp:
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.cpp:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2019-2022 Apple Inc. All rights reserved.
```
2022

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/WasmStruct.h:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2017-2022 Apple Inc. All rights reserved.
```
2022

dbezhetskov replied:
fixed

saambarati commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -322,6 +322,37 @@ WASM_SLOW_PATH_DECL(ref_func)
+
+    FixedVector<uint64_t> values(structType.fieldCount());
+    for (unsigned i = 0; i < structType.fieldCount(); ++i)
+        values[i] = READ((instruction.m_firstValue - i)).encodedJSValue();
```
We shouldn't need this intermediate FixedVector. I don't think it's buying us anything

dbezhetskov replied:
Cool, so I can pass a pointer to the first value, right?
Sorry for the newbie question, but how can I do this? :)

justinmichaud replied:
Maybe assert that instruction.m_firstValue is on the stack, and then you should be able to pass a pointer to the first value unless I am missing something

saambarati commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3067,6 +3070,104 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+    append(Move, resultStruct, result);
+
+    return { };
+}
```
This code shouldn't need to memmove all the values to the stack and then memmove them to the struct. We should just allocate a struct and store directly to it.

dbezhetskov replied:
Added a fixme comment and the follow-up bug

saambarati commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3067,6 +3070,104 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+    }
+
+    return { };
+}
```
Why not just inline this load?

saambarati commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3067,6 +3070,104 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+    emitCCall(&operationWasmStructSet, TypedTmp(), instanceValue(), structReference, addConstant(Types::I32, fieldIndex), valueAsInt64);
+
+    return { };
+}
```
We should inline this store

saambarati commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2324,6 +2327,109 @@ auto B3IRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> Pa
+    }
+
+    return { };
+}
```
We should inline this load. This can be a follow up, but we need a bug opened for this and a FIXME that links to it.

saambarati commented on Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
```diff
@@ -163,6 +163,37 @@ static inline void emitThrowImpl(CCallHelpers& jit, unsigned exceptionIndex)
     });
 }
 
+static inline void emitStructNewImpl(CCallHelpers& jit, unsigned typeIndex)
```
We should inline this allocation, either now or in a followup. If a followup, let's open a bug for it and link to it here.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2037,6 +2043,26 @@ Value* B3IRGenerator::emitAtomicCompareExchange(ExtAtomicOpType op, Type valueTy
+    case TypeKind::I64:
+    case TypeKind::F32:
+    case TypeKind::F64:
+        m_currentBlock->appendNew<MemoryValue>(m_proc, memoryKind(Store), origin(), get(argument), structValue, offset);
```
Write barrier?

dbezhetskov replied:
Added

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3067,6 +3071,85 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+        break;
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
```
Write barrier?

dbezhetskov replied:
Added

justinmichaud commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2324,6 +2350,60 @@ auto B3IRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> Pa
+{
+    const auto type = Type { TypeKind::Ref, Nullable::No, m_info.typeSignatures[typeIndex]->index() };
+
+    Value* structValue = m_currentBlock->appendNew<CCallValue>(m_proc, toB3Type(type), origin(),
```
FIXME to inline this?

dbezhetskov replied:
Added

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3067,6 +3071,91 @@ auto AirIRGenerator::addI31GetU(ExpressionType ref, ExpressionType& result) -> P
+    case TypeKind::RefNull:
+    case TypeKind::I64: {
+        if (isRefType(fieldType))
+            emitWriteBarrierForJSWrapper();
```
This colors the whole Instance, not just the target struct cell, right? We only need to color the target struct cell here,

dbezhetskov replied:
Right, but we mark a whole instance for global.set too. Weird, maybe there is a reason to do this, but I don't know it.

dbezhetskov replied:
Meanwhile I fixed struct.set implementation to use a more precise write barrier, for Air and B3.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -800,6 +804,7 @@ class AirIRGenerator {
     void emitLoopTierUpCheck(uint32_t loopIndex, const Vector<TypedTmp>& liveValues);
 
     void emitWriteBarrierForJSWrapper();
+    void emitWriteBarrier(TypedTmp cell, bool cellIsInstance = false);
```
Let's avoid using `bool`. Use `enum` for readability.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.h:
```diff
@@ -0,0 +1,59 @@
+
+#if ENABLE(WEBASSEMBLY)
+
+#include "JSDestructibleObject.h"
```
We do not need this header.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.cpp:
```diff
@@ -0,0 +1,91 @@
+    UNUSED_PARAM(callFrame);
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return JSValue::encode(throwTypeError(globalObject, scope, "WebAssembly.Struct is not accessible from JS"_s));
```
You can use `return throwVMTypeError` instead.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.cpp:
```diff
@@ -0,0 +1,91 @@
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return JSValue::encode(throwTypeError(globalObject, scope, "WebAssembly.Struct is not accessible from JS"_s));
```
Ditto.

dbezhetskov replied:
fixed

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,229 @@
+{
+}
+
+JSWebAssemblyStruct* JSWebAssemblyStruct::create(JSWebAssemblyInstance* instance, uint32_t typeIndex)
```
Since this function can fail (throwing an error), let's rename it to `tryCreate`.

Constellation replied:
And since this function throws an error, the first parameter should be `JSGlobalObject*`.

dbezhetskov replied:
This function is called in B3 and Air so adding an extra parameter will increase the code size and actually, `JSGlobalObject*` could be obtained through `Instance`. So, I would like to postpone this change before the patch that will inline this allocation in Air and B3. WDYT?

Renamed to `tryCreate`.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -182,6 +182,7 @@ constexpr bool typeExposedByDefault = true;
     macro(WebAssemblyMemory,       webAssemblyMemory,       webAssemblyMemory,       JSWebAssemblyMemory,       Memory,       object, typeExposedByDefault) \
     macro(WebAssemblyModule,       webAssemblyModule,       webAssemblyModule,       JSWebAssemblyModule,       Module,       object, typeExposedByDefault) \
     macro(WebAssemblyRuntimeError, webAssemblyRuntimeError, webAssemblyRuntimeError, ErrorInstance,             RuntimeError, error,  typeExposedByDefault) \
+    macro(WebAssemblyStruct,       webAssemblyStruct,       webAssemblyStruct,       JSWebAssemblyStruct,       Struct,       object, typeExposedByDefault) \
```
Hide if `useWebAssemblyGC` is false.

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,229 @@
+
+    ASSERT(fieldIndex < structType()->fieldCount());
+    const uint32_t* targetPointer = reinterpret_cast<const uint32_t*>(reinterpret_cast<const uint8_t*>(this) + getFieldOffset(fieldIndex, structType()));
+
```
This mechanism does not work since IsoSubspace requires all the cell has the same size. So you cannot use these tail region.

Constellation reviewed (changes requested):
Commented. This JSWebAssemblyStruct is not having data payload.

justinmichaud commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h:
```diff
@@ -0,0 +1,91 @@
+
+    FixedVector<uint8_t> m_payload;
+
+    uint8_t* m_payload_data { nullptr };
```
Why do we store this too? Shouldn't we just do m_payload.data() now and offsetOfStorage for JIT code?

dbezhetskov replied:
Ah, of course. I haven't known that `offsetOfStorage` exists :) Thanks!

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -2187,6 +2158,54 @@ void AirIRGenerator::sanitizeAtomicResult(ExtAtomicOpType op, Type valueType, Tm
+    BasicBlock* doSlowPath = m_code.addBlock();
+    BasicBlock* continuation = m_code.addBlock();
+
+    if (cellIsInstance == ShouldUseCellAsInstance::Yes)
```
I am not a big fan of this flag, it is kind of confusing. Why take in an instance tmp, and make a helper function for the cellIsInstance::no case that loads it? emitWriteBarrierForJSWrapper is the only user of that case anyway, right?

dbezhetskov replied:
Yeah, I see, I removed the flag, so now it should be less confusing: `emitBarrier` now takes two arguments `emitBarrier(cell, instance)`, in case of *jsWrapper* `cell` and `instance` are equal, but for other, more grained applications they can differ.

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+JSWebAssemblyStruct::JSWebAssemblyStruct(VM& vm, Structure* structure, Ref<Wasm::TypeDefinition>&& type)
+    : Base(vm, structure)
+    , m_type(WTFMove(type))
+    , m_payload(structPayloadSize(structType()))
```
Yes, this needs to be initialized at constructor, perfect!
Can you check it is initializing all with zeros? I think probably not if it is POD, so you need to init with memset if it is not initialized

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    switch (kind) {
+    case TypeKind::I32:
+    case TypeKind::F32: {
+        return 32;
```
Is 32 correct? Is it a bit size instead of byte size?

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    }
+    case TypeKind::I64:
+    case TypeKind::F64: {
+        return 64;
```
Ditto

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull: {
+        return sizeof(WriteBarrierBase<Unknown>);
```
If this is bit size, the this is wrong since it is byte size

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+
+namespace JSC {
+
+constexpr size_t typeKindSize(Wasm::TypeKind kind)
```
If it is bit size, name of this function should be explicit

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    using Wasm::TypeKind;
+
+    ASSERT(fieldIndex < structType()->fieldCount());
+    const uint32_t* targetPointer = reinterpret_cast<const uint32_t*>(m_payload.data() + getFieldOffset(fieldIndex, structType()));
```
Let's avoid this cast, and first getting uint8_t*

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    switch (field(fieldIndex).type.kind) {
+    case TypeKind::I32:
+    case TypeKind::F32:
+        return *targetPointer;
```
and cast here to uint32_t

Constellation replied:
And use bitwise_cast

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    case TypeKind::F32:
+        return *targetPointer;
+    case TypeKind::I64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
```
Use bitwise_cast

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    case TypeKind::I64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
+    case TypeKind::F64:
+        return *reinterpret_cast<const uint64_t*>(targetPointer);
```
Use bitwise_cast

Constellation replied:
It is not resolved yet.

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    uint8_t* targetPointer = m_payload.data() + getFieldOffset(fieldIndex, structType());
+    switch (field(fieldIndex).type.kind) {
+    case TypeKind::I32: {
+        *reinterpret_cast<int32_t*>(targetPointer) = argument.toInt32(globalObject);
```
Ditto

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+        return;
+    }
+    case TypeKind::F32: {
+        *reinterpret_cast<uint32_t*>(targetPointer) = bitwise_cast<uint32_t>(argument.toFloat(globalObject));
```
Ditto

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+        return;
+    }
+    case TypeKind::I64: {
+        *reinterpret_cast<uint64_t*>(targetPointer) = static_cast<uint64_t>(argument.toBigInt64(globalObject));
```
Ditto

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+        return;
+    }
+    case TypeKind::F64: {
+        *reinterpret_cast<uint64_t*>(targetPointer) = bitwise_cast<uint64_t>(argument.toNumber(globalObject));
```
Ditto

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,220 @@
+    auto* wasmStruct = jsCast<JSWebAssemblyStruct*>(cell);
+    for (unsigned i = 0; i < wasmStruct->structType()->fieldCount(); ++i) {
+        if (isRefType(wasmStruct->field(i).type))
+            visitor.append(bitwise_cast<WriteBarrier<Unknown>>(wasmStruct->get(i)));
```
This is wrong since it is passing a value, not reference. Need to get WriteBarrier* and pass the reference.

Constellation reviewed (commented):
Commented

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -1775,44 +1782,8 @@ auto AirIRGenerator::setGlobal(uint32_t index, ExpressionType value) -> PartialR
-    append(Jump);
-    m_currentBlock->setSuccessors(continuation);
-    m_currentBlock = continuation;
+    emitWriteBarrier(cell, cell);
```
This looks incorrect since `emitWriteBarrier` is scratching `instanceCell` second parameter.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -2220,6 +2191,48 @@ void AirIRGenerator::sanitizeAtomicResult(ExtAtomicOpType op, Type valueType, Tm
+    BasicBlock* doSlowPath = m_code.addBlock();
+    BasicBlock* continuation = m_code.addBlock();
+
+    append(Move, Arg::addr(instanceValue(), Instance::offsetOfOwner()), instanceCell);
```
Why do we need this? The parameter `instanceCell` should already have instance, right?

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp:
```diff
@@ -0,0 +1,229 @@
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSizeInBytes(type->field(i).type.kind);
+    return fieldOffset;
+}
```
Computing this every time accessing field is very costly. We should put this offset information to StructType.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3386,6 +3397,104 @@ auto AirIRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result
+    append(Move, Arg::addr(instanceValue(), Instance::offsetOfOwner()), jsInstance);
+    // FIXME: inline the allocation.
+    // https://bugs.webkit.org/show_bug.cgi?id=244388
+    emitCCall(&JSWebAssemblyStruct::tryCreate, result, jsInstance, addConstant(Types::I32, typeIndex));
```
You cannot call a function which is not defined with `JSC_DECLARE_JIT_OPERATION`. So you need to use operationWasmStructNew, or define a function in WasmOperations.cpp.

Constellation commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -120,9 +118,15 @@ class StructType {
 private:
     FieldType* m_payload;
     StructFieldCount m_fieldCount;
+    FixedVector<size_t> m_fieldOffsets;
```
I think, we should validate that the total size of the struct does not exceed unsigned (if it exceeds, then anyway, it cannot be allocated in wasm due to 4GB memory), and use `unsigned` offset here.

Constellation commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -98,6 +132,20 @@ void StructType::dump(PrintStream& out) const
+        currentFieldOffset += typeKindSizeInBytes(field(fieldIndex).type.kind);
+    }
+
+    m_instancePayloadSize = WTF::roundUpToMultipleOf<sizeof(uint64_t)>(currentFieldOffset.value());
```
This is not what we would like to have. We should fail the parsing, not crashing.

Constellation replied:
We need to ensure that,
1. FunctionParser checks this overflow and fail the parsing
2. We should have a test for that.

dbezhetskov replied:
1. I guess it is not a responsibility of FunctionParser, this check should be done when we create struct type, in WasmSectionParser, argee? Because we don't want to have type with invalid offsets, right?
2. Test is added, but because of our hard limit for max number of struct fields it throws different message.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2509,6 +2549,64 @@ auto B3IRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result)
+auto B3IRGenerator::addStructGet(ExpressionType structReference, const StructType& structType, uint32_t fieldIndex, ExpressionType& result) -> PartialResult
+{
+    Value* structValue = get(structReference);
+    int32_t payloadFieldOffset = fixupPointerPlusOffset(structValue, JSWebAssemblyStruct::offsetOfPayload());
```
We do not need this. `offsetOfPayload()` is constant and small. Rather, considering about the possibility of rewritten value of structValue makes the code hard to read.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2509,6 +2549,64 @@ auto B3IRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result)
+{
+    Value* structValue = get(structReference);
+    int32_t payloadFieldOffset = fixupPointerPlusOffset(structValue, JSWebAssemblyStruct::offsetOfPayload());
+    Value* payloadBase = m_currentBlock->appendNew<MemoryValue>(m_proc, memoryKind(Load), pointerType(), origin(), structValue, payloadFieldOffset);
```
Let's directly use `JSWebAssemblyStruct::offsetOfPayload()` for `payloadFieldOffset`.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2509,6 +2549,64 @@ auto B3IRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result)
+    int32_t payloadFieldOffset = fixupPointerPlusOffset(structValue, JSWebAssemblyStruct::offsetOfPayload());
+    Value* payloadBase = m_currentBlock->appendNew<MemoryValue>(m_proc, memoryKind(Load), pointerType(), origin(), structValue, payloadFieldOffset);
+
+    int32_t fieldOffset = fixupPointerPlusOffset(payloadBase, *structType.getFieldOffset(fieldIndex));
```
This makes sense.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2043,6 +2055,34 @@ Value* B3IRGenerator::emitAtomicCompareExchange(ExtAtomicOpType op, Type valueTy
+void B3IRGenerator::emitStructSet(Value* structValue, uint32_t fieldIndex, const StructType& structType, ExpressionType& argument)
+{
+    int32_t payloadFieldOffset = fixupPointerPlusOffset(structValue, JSWebAssemblyStruct::offsetOfPayload());
+    Value* payloadBase = m_currentBlock->appendNew<MemoryValue>(m_proc, memoryKind(Load), Int64, origin(), structValue, payloadFieldOffset);
```
Ditto. Let's just use `JSWebAssemblyStruct::offsetOfPayload()`.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -2043,6 +2055,34 @@ Value* B3IRGenerator::emitAtomicCompareExchange(ExtAtomicOpType op, Type valueTy
+    int32_t payloadFieldOffset = fixupPointerPlusOffset(structValue, JSWebAssemblyStruct::offsetOfPayload());
+    Value* payloadBase = m_currentBlock->appendNew<MemoryValue>(m_proc, memoryKind(Load), Int64, origin(), structValue, payloadFieldOffset);
+
+    int32_t fieldOffset = fixupPointerPlusOffset(payloadBase, *structType.getFieldOffset(fieldIndex));
```
This makes sense.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3390,6 +3401,102 @@ auto AirIRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result
+
+auto AirIRGenerator::addStructSet(ExpressionType structReference, const StructType& structType, uint32_t fieldIndex, ExpressionType value) -> PartialResult
+{
+    uint32_t payloadOffset = fixupPointerPlusOffset(structReference, JSWebAssemblyStruct::offsetOfPayload());
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -3390,6 +3401,102 @@ auto AirIRGenerator::addArrayLen(ExpressionType arrayref, ExpressionType& result
+
+auto AirIRGenerator::addStructGet(ExpressionType structReference, const StructType& structType, uint32_t fieldIndex, ExpressionType& result) -> PartialResult
+{
+    uint32_t payloadOffset = fixupPointerPlusOffset(structReference, JSWebAssemblyStruct::offsetOfPayload());
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

Constellation reviewed (approved):
r=me

dbezhetskov committed:
Support struct.new, struct.get and struct.set.
https://bugs.webkit.org/show_bug.cgi?id=243208

Reviewed by Yusuke Suzuki.

This patch introduces basic support of all main structs operations - struct.new, struct.get and struct.set
from the Wasm GC proposal (https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#structures).
In this patch struct is represented as a contiguous memory space with a corresponding JS part.
All operations are implemented via runtime calls so it's not optimal for B3 and Air tiers
but optimization (replacing runtime call to a read/write by known offset) for struct.get/set will be added as a next step.

* JSTests/wasm/gc/structs.js:
(testStructDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources-output.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::AirIRGenerator::emitWriteBarrier):
(JSC::Wasm::AirIRGenerator::addStructNew):
(JSC::Wasm::AirIRGenerator::addStructGet):
(JSC::Wasm::AirIRGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitWriteBarrierForJSWrapper):
(JSC::Wasm::B3IRGenerator::emitWriteBarrier):
(JSC::Wasm::B3IRGenerator::emitStructSet):
(JSC::Wasm::B3IRGenerator::addStructNew):
(JSC::Wasm::B3IRGenerator::addStructGet):
(JSC::Wasm::B3IRGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseStructTypeIndex):
(JSC::Wasm::FunctionParser<Context>::parseStructFieldIndex):
(JSC::Wasm::FunctionParser<Context>::parseStructTypeIndexAndFieldIndex):
(JSC::Wasm::FunctionParser<Context>::parseStructFieldManipulation):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::addStructNew):
(JSC::Wasm::LLIntGenerator::addStructGet):
(JSC::Wasm::LLIntGenerator::addStructSet):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseStructType):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::StructType::StructType):
(JSC::Wasm::TypeDefinition::tryCreateStructType):
(JSC::Wasm::StructParameterTypes::translate):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::typeKindSizeInBytes):
(JSC::Wasm::StructType::getFieldOffset const):
(JSC::Wasm::StructType::getFieldOffset):
(JSC::Wasm::StructType::instancePayloadSize const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedStructSize):
(JSC::Wasm::StructType::StructType): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyHelpers.h:
(JSC::toJSValue):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp: Added.
(JSC::JSWebAssemblyStruct::JSWebAssemblyStruct):
(JSC::JSWebAssemblyStruct::tryCreate):
(JSC::JSWebAssemblyStruct::fieldPointer const):
(JSC::JSWebAssemblyStruct::fieldPointer):
(JSC::JSWebAssemblyStruct::get const):
(JSC::JSWebAssemblyStruct::set):
(JSC::JSWebAssemblyStruct::finishCreation):
(JSC::JSWebAssemblyStruct::visitChildrenImpl):
(JSC::JSWebAssemblyStruct::destroy):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.h: Added.
* Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.cpp: Added.
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::WebAssemblyStructConstructor::create):
(JSC::WebAssemblyStructConstructor::createStructure):
(JSC::WebAssemblyStructConstructor::finishCreation):
(JSC::WebAssemblyStructConstructor::WebAssemblyStructConstructor):
* Source/JavaScriptCore/wasm/js/WebAssemblyStructConstructor.h: Added.
* Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.cpp: Added.
(JSC::WebAssemblyStructPrototype::create):
(JSC::WebAssemblyStructPrototype::createStructure):
(JSC::WebAssemblyStructPrototype::finishCreation):
(JSC::WebAssemblyStructPrototype::WebAssemblyStructPrototype):
* Source/JavaScriptCore/wasm/js/WebAssemblyStructPrototype.h: Added.
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/255586@main
